### PR TITLE
Scope the appdata theming storage for global and users

### LINKED ISSUE #34596: [Bug] [25.0.0RC4] Custom background doesn't stay (404)
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on Github _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] Nextcloud Server **is** running on 64bit capable CPU, PHP and OS.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

Getting a 404 on a custom background: `GET https://.../index.php/apps/theming/background`

```
MariaDB [nextcloud]> select * from oc_preferences where userid='alice' and appid in ('dashboard', 'theming');
+--------+-----------+-------------------+-------------+
| userid | appid     | configkey         | configvalue |
+--------+-----------+-------------------+-------------+
| alice  | dashboard | firstRun          | 0           |
| alice  | theming   | background        | custom      |
| alice  | theming   | backgroundVersion | 1           |
+--------+-----------+-------------------+-------------+
3 rows in set (0.001 sec)
```

file "background.jpg" is missing on disk in "data/appdata_*/theming/alice"

### Steps to reproduce

1. Create three users Alice, Bob and Charles
2. Login as Alice
3. Upload an image
4. Set that image as custom background
5. Go to files app and shift+refresh to see it

### Expected behavior

Background appears and exists on disk

### Installation method

_No response_

### Operating system

_No response_

### PHP engine version

_No response_

### Web server

_No response_

### Database engine version

_No response_

### Is this bug present after an update or on a fresh install?

_No response_

### Are you using the Nextcloud Server Encryption module?

_No response_

### What user-backends are you using?

- [ ] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

_No response_

### List of activated Apps

```shell
N/A
```

### Nextcloud Signing status

_No response_

### Nextcloud Logs

```shell
{
  "reqId": "yGU73Yn4alZXEjEKqNf7",
  "level": 3,
  "time": "2022-10-14T07:50:24+00:00",
  "remoteAddr": "127.0.0.1",
  "user": "--",
  "app": "index",
  "method": "GET",
  "url": "/index.php/apps/theming/background?v=2",
  "message": "Call to a member function getUID() on null",
  "userAgent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/106.0.0.0 Safari/537.36",
  "version": "25.0.0.16",
  "exception": {
    "Exception": "Error",
    "Message": "Call to a member function getUID() on null",
    "Code": 0,
    "Trace": [
      {
        "function": "__construct",
        "class": "OCA\\Theming\\Controller\\UserThemeController",
        "type": "->",
        "args": [
          "theming",
          {
            "__class__": "OC\\AppFramework\\Http\\Request"
          },
          {
            "__class__": "OC\\AllConfig"
          },
          {
            "__class__": "OC\\User\\Session"
          },
          {
            "__class__": "OCA\\Theming\\Service\\ThemesService"
          },
          {
            "__class__": "OCA\\Theming\\ThemingDefaults"
          },
          {
            "__class__": "OCA\\Theming\\Service\\BackgroundService"
          }
        ]
      },
      {
        "file": "/srv/www/htdocs/server/lib/private/AppFramework/Utility/SimpleContainer.php",
        "line": 108,
        "function": "newInstanceArgs",
        "class": "ReflectionClass",
        "type": "->",
        "args": [
          [
            "theming",
            {
              "__class__": "OC\\AppFramework\\Http\\Request"
            },
            {
              "__class__": "OC\\AllConfig"
            },
            {
              "__class__": "OC\\User\\Session"
            },
            {
              "__class__": "OCA\\Theming\\Service\\ThemesService"
            },
            {
              "__class__": "OCA\\Theming\\ThemingDefaults"
            },
            {
              "__class__": "OCA\\Theming\\Service\\BackgroundService"
            }
          ]
        ]
      },
      {
        "file": "/srv/www/htdocs/server/lib/private/AppFramework/Utility/SimpleContainer.php",
        "line": 116,
        "function": "buildClass",
        "class": "OC\\AppFramework\\Utility\\SimpleContainer",
        "type": "->",
        "args": [
          {
            "__class__": "ReflectionClass",
            "name": "OCA\\Theming\\Controller\\UserThemeController"
          }
        ]
      },
      {
        "file": "/srv/www/htdocs/server/lib/private/AppFramework/Utility/SimpleContainer.php",
        "line": 133,
        "function": "resolve",
        "class": "OC\\AppFramework\\Utility\\SimpleContainer",
        "type": "->",
        "args": [
          "OCA\\Theming\\Controller\\UserThemeController"
        ]
      },
      {
        "file": "/srv/www/htdocs/server/lib/private/AppFramework/DependencyInjection/DIContainer.php",
        "line": 465,
        "function": "query",
        "class": "OC\\AppFramework\\Utility\\SimpleContainer",
        "type": "->",
        "args": [
          "OCA\\Theming\\Controller\\UserThemeController"
        ]
      },
      {
        "file": "/srv/www/htdocs/server/lib/private/AppFramework/DependencyInjection/DIContainer.php",
        "line": 437,
        "function": "queryNoFallback",
        "class": "OC\\AppFramework\\DependencyInjection\\DIContainer",
        "type": "->",
        "args": [
          "OCA\\Theming\\Controller\\UserThemeController"
        ]
      },
      {
        "file": "/srv/www/htdocs/server/lib/private/AppFramework/App.php",
        "line": 159,
        "function": "query",
        "class": "OC\\AppFramework\\DependencyInjection\\DIContainer",
        "type": "->",
        "args": [
          "OCA\\Theming\\Controller\\UserThemeController"
        ]
      },
      {
        "file": "/srv/www/htdocs/server/lib/private/Route/Router.php",
        "line": 298,
        "function": "main",
        "class": "OC\\AppFramework\\App",
        "type": "::",
        "args": [
          "OCA\\Theming\\Controller\\UserThemeController",
          "getBackground",
          {
            "__class__": "OC\\AppFramework\\DependencyInjection\\DIContainer"
          },
          [
            "theming.userTheme.getBackground"
          ]
        ]
      },
      {
        "file": "/srv/www/htdocs/server/lib/base.php",
        "line": 1047,
        "function": "match",
        "class": "OC\\Route\\Router",
        "type": "->",
        "args": [
          "/apps/theming/background"
        ]
      },
      {
        "file": "/srv/www/htdocs/server/index.php",
        "line": 36,
        "function": "handleRequest",
        "class": "OC",
        "type": "::",
        "args": []
      }
    ],
    "File": "/srv/www/htdocs/server/apps/theming/lib/Controller/UserThemeController.php",
    "Line": 76,
    "CustomMessage": "--"
  }
}
```

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by skjnldsv**
Fix https://github.com/nextcloud/server/issues/34596

### CODE REVIEW

**Comment by PVince81** [comment]
- [ ] add hash to the "v=" argument based on user id to avoid seeing another user's background
- [ ] Simon's problem: still need shift+refresh for seeing the new background, might need to also cache bust the CSS file containing the variables: also append the backgroundVersion to the URL of that CSS file ?!

**Review by PVince81 — APPROVED** [review]
:+1: for @skjnldsv's part, please also review what I pushed

**Comment by PVince81** [comment]
I've [pushed a fix](https://github.com/nextcloud/server/pull/34599/commits/8a5b5263d8439249f3bdcd580231f5deb1124ac1) that adds cache busting also for the CSS variables to make sure we load the correct background in apps.
Additionally I've hashed the user id there to make sure we don't see the cached background of another user.
This fixes https://github.com/nextcloud/server/pull/34599#issuecomment-1278792810

File: apps/theming/lib/Themes/DefaultTheme.php [hunk 1]
```diff
@@ -242,7 +242,8 @@ public function getCSSVariables(): array {
 			$currentVersion = (int)$this->config->getUserValue($user->getUID(), Application::APP_ID, 'backgroundVersion', '0');
 
 			if ($themingBackground === 'custom') {
-				$variables['--image-main-background'] = "url('" . $this->urlGenerator->linkToRouteAbsolute('theming.userTheme.getBackground') . "?v=$currentVersion')";
+				$cacheBuster = sha1($user->getUID() . '_' . $currentVersion);
+				$variables['--image-main-background'] = "url('" . $this->urlGenerator->linkToRouteAbsolute('theming.userTheme.getBackground') . "?v=$cacheBuster')";
```
Comment (szaimen) [inline apps/theming/lib/Themes/DefaultTheme.php:246]: I guess we need to increase the version also when admin theming changes in order to fix public pages?

Comment (skjnldsv) [reply apps/theming/lib/Themes/DefaultTheme.php:246]: No, this is only about user theming here.
For the admin theming this is done already here: https://github.com/nextcloud/server/blob/1849ac70ea7ad4515bd6a2d88b193ab076ba2f40/apps/theming/lib/ImageManager.php#L81

Comment (skjnldsv) [reply apps/theming/lib/Themes/DefaultTheme.php:246]: Admin:
https://github.com/nextcloud/server/blob/1849ac70ea7ad4515bd6a2d88b193ab076ba2f40/apps/theming/lib/Themes/DefaultTheme.php#L219-L232

User: 
https://github.com/nextcloud/server/blob/1849ac70ea7ad4515bd6a2d88b193ab076ba2f40/apps/theming/lib/Themes/DefaultTheme.php#L238-L252

Comment (szaimen) [reply apps/theming/lib/Themes/DefaultTheme.php:246]: All right! Then I don't understand why admin themin changes are not applied to public pages without force refresh...

**Comment by szaimen** [comment]
@skjnldsv Then I don't understand why admin themin changes are not applied to public pages without force refresh...

**Comment by skjnldsv** [comment]
> @skjnldsv Then I don't understand why admin themin changes are not applied to public pages without force refresh...

Ah, because the css needs to be bump too :see_no_evil: 
https://github.com/nextcloud/server/blob/8a5b5263d8439249f3bdcd580231f5deb1124ac1/apps/theming/lib/Service/ThemeInjectionService.php#L90-L102

We only do it for users. Shall I handle it here or followup?

**Review by szaimen — CHANGES_REQUESTED** [review]
See 

[resolved](https://user-images.githubusercontent.com/42591237/195839791-94c1f8da-5e62-4fdb-9288-79fdb31d1073.mp4)

**Comment by szaimen** [comment]
Also, shouldn't the color of the buttons change to the primary color when you change the theme?
image

**Comment by szaimen** [comment]
As last thing, this button reset to the by the user chosen primary color and not to nextcloud default blue:
image

And as soon as you choose a custom color does the background picture get removed:

https://user-images.githubusercontent.com/42591237/195840826-84491359-af4a-485c-bb36-e8c942de5df4.mp4

**Comment by PVince81** [comment]
> We only do it for users. Shall I handle it here or followup?

how would you do it ? probably need another cache buster value ? there seems to be one in the admin theme as we saw it on disk

**Comment by skjnldsv** [comment]
Pushing fix in a second

**Review by szaimen — APPROVED** [review]
works but didn't review the code

**Comment by PVince81** [comment]
I've retested admin vs user background and no more shift+refresh is needed, it works smoothly and as expected.
Likewise, the public page had the background configured by the admin

Shipit!

**Comment by PVince81** [comment]
I've raised the lower priority https://github.com/nextcloud/server/issues/34606 to clean up the old theme cache data since we now moved everything into "appdata*/theming/global/"

**Comment by blizzz** [comment]
`backgroundVersion` is not used anymore, but still migrated in `core/Migrations/Version25000Date20221007010957.php`. Can be cleaned up afterwards though, just not to forget it.

**Pull request merged**